docs/contributing/Feature-flags.md

```diff
@@ -0,0 +1,328 @@
+
+   ```dart
+   const Feature unicornEmojis = Feature(
+     name: 'add unicorn emojis in lots of fun places',
+     runtimeId: 'enable-unicorn-emojis',
+   );
+   ```
+
+   To allow an environment variable, include `environmentOverride`:
+
+   ```dart
+   const Feature unicornEmojis = Feature(
+     name: 'add unicorn emojis in lots of fun places',
+     environmentOverride: 'FLUTTER_UNICORN_EMOJIS',
+   );
+   ```
+
+1. Add a new field to `abstract class FeatureFlags`:
+
+   ```dart
+   abstract class FeatureFlags {
+     /// Whether to add unicorm emojis in lots of fun places.
+     bool get isUnicornEmojisEnabled => false;
+   }
+   ```
+
+1. Add an implement the same getter in [`FlutterFeatureFlagsIsEnabled`][]:
+
+   ```dart
+   mixin FlutterFeatureFlagsIsEnabled implements FeatureFlags {
+     @override
+     bool get isUnicornEmojisEnabled => isEnabled(unicornEmojis);
+   }
+   ```
+
+   [`FlutterFeatureFlagsIsEnabled`]: ../../packages/flutter_tools/lib/src/flutter_features.dart
+
+### Allowing flags to be enabled
+
+By default, after [adding a flag](#adding-a-flag), the flag is considered
+_disabled_, and _cannot_ be enabled outside of our own [unit tests](#tests).
+This allows iterating locally with the code without having to support users or
+field issues related to the flag.
+
+After some time, you may want to allow the flag to be enabled.
+
+Using the options `master`, `beta` or `stable`, you can make the flag
+configurable in those channels. For example, to make the flag available to be
+enabled (but still off by default) on the `master` channel:
+
+```dart
+const Feature unicornEmojis = Feature(
+  name: 'add unicorn emojis in lots of fun places',
+  configSetting: 'enable-unicorn-emojis',
+  master: FeatureChannelSetting(available: true),
+);
+```
+
+Or to make it available on all channels:
+
+```dart
+const Feature unicornEmojis = Feature(
+  name: 'add unicorn emojis in lots of fun places',
+  configSetting: 'enable-unicorn-emojis',
+  master: FeatureChannelSetting(available: true),
+  beta: FeatureChannelSetting(available: true),
+  stable: FeatureChannelSetting(available: true),
+);
+```
+
+### Enabling a flag by default
+
+Once a flag is ready to be enabled by default, once again it can be configured
+on a per-channel basis.
+
+For example, enabled on `master` by default, but disabled elsewhere:
+
+```dart
+const Feature unicornEmojis = Feature(
+  name: 'add unicorn emojis in lots of fun places',
+  configSetting: 'enable-unicorn-emojis',
+  master: FeatureChannelSetting(available: true, enabledByDefault: true),
+  beta: FeatureChannelSetting(available: true),
+  stable: FeatureChannelSetting(available: true),
+);
+```
+
+Once the flag is ready to be enabled in every environment:
+
+```dart
+const Feature unicornEmojis = Feature.fullyEnabled(
+  name: 'add unicorn emojis in lots of fun places',
+  configSetting: 'enable-unicorn-emojis',
+);
+```
+
+### Removing a flag
```

Comment:
Do we want to have guidance around the lifetime of the flag:
1. File an issue describing the flagged experiment
2. Track when the flags are getting promoted (maybe labels that we could automate)
3. Track when a flag should be removed

Its also a place for any reports to link back to if they are filed.